measure

```diff
@@ -218,44 +204,32 @@ class SignalFx(Measure):
                     'Failed to query SignalFx for program {}. Error: {}'.format(
                     program, str(e)))
 
+            metadata_objects = {}
+            data_messages = []
+
             # process messages in computation stream
             for msg in computation.stream():
+                if isinstance(msg, signalfx.signalflow.messages.MetadataMessage):
+                    metadata_objects[msg.tsid] = msg.properties
                 if isinstance(msg, signalfx.signalflow.messages.DataMessage):
-                    space_vals = [x for x in list(msg.data.values()) if isinstance(
-                        x, numbers.Number)]
-                    if len(space_vals) == 0:
-                        n_nodata_msg += 1
-                    else:
-                        n_data_msg += 1
-                        space_aggr_vals.append(self._aggregate_values(space_vals,
-                            space_aggr))
-
-        # verify percent of empty data messages is below configured threshold
-        n_msg = n_data_msg + n_nodata_msg
-        nodata_pct = 100.0 * (n_nodata_msg / n_msg)
-        self._nfy('SignalFlow stream: {} data msgs, {} nodata msgs, {:.2f}% empty for program {}'.format(
-            n_data_msg, n_nodata_msg, nodata_pct, program))
-        assert nodata_pct <= nodata_pct_ok, \
-            'SignalFlow stream:  {:.2f}% of {} data msgs are empty ({:.2f}% is ok) for program {}'.format(
-            nodata_pct, n_msg, nodata_pct_ok, program)
-        assert len(space_aggr_vals) > 0, \
-            'SignalFx query for program {} returned no time series metrics data'.format(
-            program)
-
-        # aggregate in time and return value
-        return self._aggregate_values(space_aggr_vals, time_aggr)
-
-    # helper:  return an aggregate value computed from an input list of values
-    def _aggregate_values(self, vals, aggr):
-        if aggr == 'avg':
-            return sum(vals) / float(len(vals))
-        if aggr == 'max':
-            return max(vals)
-        if aggr == 'min':
-            return min(vals)
-        if aggr == 'sum':
-            return sum(vals)
-        raise Exception('Unexpected aggregation method {}'.format(aggr))
+                    data_messages.append(msg)
+
+            # Read more about 'sf_key' at https://developers.signalfx.com/signalflow_analytics/rest_api_messages/stream_messages_specification.html#_metadata_messages
+            inapplicable_sf_keys = ['sf_metric', 'sf_originatingMetric', 'computationId']
+            inst_ret = OrderedDict()
+
+            for tsid, props in metadata_objects.items():
+                group_id_key = next(filter(lambda key: key not in inapplicable_sf_keys, props['sf_key']))
+                group_id = props[group_id_key]
+                inst_ret[tsid] = {'id': str(group_id), 'data': []}
+
+            for msg in data_messages:
```

Comment:
Add comment to this code block, e.g., what it is doing, and preferably what is the expected data structure of value (what can be expected to be sent to the backend).